pkg/handlers/generic/mutation/encryption/inject.go

```diff
@@ -0,0 +1,307 @@
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/yaml"
+
+	carenv1 "github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/api/v1alpha1"
+	"github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/common/pkg/capi/clustertopology/handlers/mutation"
+	"github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/common/pkg/capi/clustertopology/patches"
+	"github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/common/pkg/capi/clustertopology/patches/selectors"
+	"github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/common/pkg/capi/clustertopology/variables"
+	"github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/common/pkg/capi/utils"
+	k8sClientUtil "github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/common/pkg/k8s/client"
+	"github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/pkg/handlers/generic/clusterconfig"
+)
+
+const (
+	// VariableName is the external patch variable name.
+	VariableName                        = "encryptionAtRest"
+	SecretKeyForEtcdEncryption          = "config"
+	defaultEncryptionSecretNameTemplate = "%s-encryption-config" //nolint:gosec // Does not contain hard coded credentials.
+	encryptionConfigurationOnRemote     = "/etc/kubernetes/pki/encryptionconfig.yaml"
+	apiServerEncryptionConfigArg        = "encryption-provider-config"
+)
+
+type encryptionPatchHandler struct {
+	client            ctrlclient.Client
+	keyGenerator      TokenGenerator
+	variableName      string
+	variableFieldPath []string
+}
+
+func NewPatch(client ctrlclient.Client, keyGenerator TokenGenerator) *encryptionPatchHandler {
+	return &encryptionPatchHandler{
+		client:            client,
+		keyGenerator:      keyGenerator,
+		variableName:      clusterconfig.MetaVariableName,
+		variableFieldPath: []string{VariableName},
+	}
+}
+
+func (h *encryptionPatchHandler) Mutate(
+	ctx context.Context,
+	obj *unstructured.Unstructured,
+	vars map[string]apiextensionsv1.JSON,
+	holderRef runtimehooksv1.HolderReference,
+	clusterKey ctrlclient.ObjectKey,
+	clusterGetter mutation.ClusterGetter,
+) error {
+	log := ctrl.LoggerFrom(ctx, "holderRef", holderRef)
+
+	encryptionVariable, err := variables.Get[carenv1.EncryptionAtRest](
+		vars,
+		h.variableName,
+		h.variableFieldPath...,
+	)
+	if err != nil {
+		if variables.IsNotFoundError(err) {
+			log.V(5).Info("encryption variable not defined")
+			return nil
+		}
+		return err
+	}
+
+	log = log.WithValues(
+		"variableName",
+		h.variableName,
+		"variableFieldPath",
+		h.variableFieldPath,
+		"variableValue",
+		encryptionVariable,
+	)
+
+	return patches.MutateIfApplicable(
+		obj, vars, &holderRef, selectors.ControlPlane(), log,
+		func(obj *controlplanev1.KubeadmControlPlaneTemplate) error {
+			cluster, err := clusterGetter(ctx)
+			if err != nil {
+				log.Error(err, "failed to get cluster from encryption mutation handler")
+				return err
+			}
```

Comment:
nit: in the other mutators this function is called early outside of the `MutateIfApplicable`, maybe we can do the same here?

Reply:
This came up during pairing with Daniel when reviewing other PRs. The mutation handler is called for all the objects so clusterGetter will be called multiple times. Even though it is cached, I think we should only make Get call if the `obj` is the target object. i. .e in `MutateIfApplicable` method. 
If you agree, we want to change the same for other handlers.

Reply:
sg, we can follow up in a separate PR for the other handlers.